org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/matching/JavaSearchNameEnvironment.java

```diff
@@ -563,4 +573,15 @@ private static boolean isComplianceJava9OrHigher(IJavaProject javaProject) {
 	}
 	return CompilerOptions.versionToJdkLevel(javaProject.getOption(JavaCore.COMPILER_COMPLIANCE, true)) >= ClassFileConstants.JDK9;
 }
+
+private static boolean hasModulePathContainer(JavaProject javaProject) throws JavaModelException {
+	IClasspathEntry[] classpath = javaProject.getRawClasspath();
+	for (IClasspathEntry entry : classpath) {
+		IPath entryPath = entry.getPath();
+		if (entryPath != null && JavaCore.MODULE_PATH_CONTAINER_ID.equals(entryPath.toString())) {
```

Comment:
This sounds like a hack on top of another hack. Would be great to understand what that `MODULE_PATH_CONTAINER` is good for and if that should be better properly configured instead. I believe at the time it was introduced there were no "module" classpath attribute.

Reply:
And? Projects like that will exist, same as during the tests. Just because its a hack doesn't mean its not being used.

Reply:
Simeon, you didn't get it. I mean, this container looks like a hack for JDT tests, nothing more. It will be NOT exist in any code outside JDT.

Reply:
Hi @mpalat ,

can you clarify the purpose and usage of `JavaCore.MODULE_PATH_CONTAINER_ID`/`ModulePathContainer`?

Was this some hack to test the compiler module path with Eclipse projects? Can we expect that projects having a classpath entry as follows, are only used in tests and don't exist as actual projects in "real" workspaces?

```
<classpathentry kind="con" path="org.eclipse.jdt.MODULE_PATH"/>
```

If this was added for tests only, we can adjust the tests that have problems with the change (to have a proper module path).

Best regards and thanks,
Simeon

Reply:
I've prepared a 2nd branch with how tests would look without using `JavaCore.MODULE_PATH_CONTAINER_ID`:

https://github.com/trancexpress/eclipse.jdt.core/commit/404f642d03eb2f1e32013552867a24d595db176a

Its not a small change.

Reply:
> Its not a small change.

In that case we should separate the change with the actual fix from the patch that changes the usage of "MODULE_PATH" container.

Regarding https://github.com/trancexpress/eclipse.jdt.core/commit/404f642d03eb2f1e32013552867a24d595db176a : I'm not sure I can follow. The "MODULE_PATH" seem to be a shortcut to include all required modules of a project module, the patch I see in the branch seem to simply add the "module" attribute to all classpath entries, not the required modules. `org.eclipse.jdt.internal.core.ModulePathContainer.getClasspathEntries()`.

So I don't understand what exactly breaks with the original patch in "MODULE_PATH" and how the second patch supposed to fix that.

Would be nice first to explain the fail and discuss the right solution after that?

Reply:
Opened #717 to check if/why we need this entry in tests.

> Would be nice first to explain the fail and discuss the right solution after that?

The new fails are due to expecting the `module` attribute for classpath entries on the module path, in order to add module information during the search. The tests that fail require having classpath entries on the module path, but don't explicitly make those entries modular. This worked, since the search environment adds module infos for every classpath entry (assuming JLS > 9). Now it no longer works.